art/attacks/evasion/imperceptible_asr/imperceptible_asr.py

```diff
@@ -0,0 +1,621 @@
+        """
+        import tensorflow.compat.v1 as tf1
+
+        # disable eager execution as Lingvo uses tensorflow.compat.v1 API
+        tf1.disable_eager_execution()
+
+        # Super initialization
+        super().__init__(estimator=estimator)
+        self.masker = masker
+        self.eps = eps
+        self.learning_rate_1 = learning_rate_1
+        self.max_iter_1 = max_iter_1
+        self.alpha = alpha
+        self.learning_rate_2 = learning_rate_2
+        self.max_iter_2 = max_iter_2
+        self._targeted = True
+        self.batch_size = batch_size
+        self._check_params()
+
+        # init some aliases
+        self._window_size = masker.window_size
+        self._hop_size = masker.hop_size
+        self._sample_rate = masker.sample_rate
+
+        # TensorFlow placeholders
+        self._delta = tf1.placeholder(tf1.float32, shape=[None, None], name="art_delta")
+        self._power_spectral_density_maximum_tf = tf1.placeholder(tf1.float32, shape=[None, None], name="art_psd_max")
+        self._masking_threshold_tf = tf1.placeholder(
+            tf1.float32, shape=[None, None, None], name="art_masking_threshold"
+        )
+        # TensorFlow loss gradient ops
+        self._loss_gradient_masking_threshold_op_tf = self._loss_gradient_masking_threshold_tf(
+            self._delta, self._power_spectral_density_maximum_tf, self._masking_threshold_tf
+        )
+
+    def generate(self, x: np.ndarray, y: np.ndarray, **kwargs) -> np.ndarray:
+        """
+        Generate imperceptible, adversarial examples.
+
+        :param x: An array with the original inputs to be attacked.
+        :param y: Target values of shape (batch_size,). Each sample in `y` is a string and it may possess different
+            lengths. A possible example of `y` could be: `y = np.array(['SIXTY ONE', 'HELLO'])`.
+        :return: An array holding the adversarial examples.
+        """
+        nb_samples = x.shape[0]
+
+        x_imperceptible = [None] * nb_samples
+
+        nb_batches = int(np.ceil(nb_samples / float(self.batch_size)))
+        for m in range(nb_batches):
+            # batch indices
+            begin, end = m * self.batch_size, min((m + 1) * self.batch_size, nb_samples)
+
+            # create batch of adversarial examples
+            x_imperceptible[begin:end] = self._generate_batch(x[begin:end], y[begin:end])
+        return np.array(x_imperceptible, dtype=object)
+
+    def _generate_batch(self, x: np.ndarray, y: np.ndarray) -> np.ndarray:
+        """Create imperceptible, adversarial sample.
+
+        This is a helper method that calls the methods to create an adversarial
+        (:func:`~art.attack.evasion.imperceptible_asr.imperceptible_asr.ImperceptibleAsr._create_adversarial`) and
+        imperceptible
+        (:func:`~art.attack.evasion.imperceptible_asr.imperceptible_asr.ImperceptibleAsr._create_imperceptible`) example
+        subsequently.
```

Comment:
Did you check if the docs build with sphinx is recognising the function syntax correctly?